# World_Bank_Data_-Wrangling - 
## Este notebook utiliza a biblioteca pandas para realizar o tratamento de uma base de dados

### Site:       https://filipedeabreu.com

### Autor:      Filipe de Abreu

### Manutenção:  Filipe de Abreu

<hr>

# Testado em
### OS: Windows 11                                      
### Versão do python: 3.12.7    

In [5]:
#importações

import pandas as pd

In [6]:
df = pd.read_excel('datasets/WDI World Bank.xlsx')

df.head(5)

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name        Series Code  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS   
1  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.RU.ZS   
2  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.UR.ZS   
3            Access to electricity (% of population)     EG.ELC.ACCS.ZS   
4  Access to electricity, rural (% of rural popul...  EG.ELC.ACCS.RU.ZS   

  2021 [YR2021]                                 Topic  
0            ..  Environment: Energy production & use  
1            ..  Environment: Energy production & use  
2            ..  Environment: Energy production & use  
3            ..  Environment: Energy production & use  
4            ..  Environment: Energy production & use

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 383577 entries, 0 to 383576
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   Country Name   383574 non-null  object
 1   Country Code   383572 non-null  object
 2   Series Name    383572 non-null  object
 3   Series Code    383572 non-null  object
 4   2021 [YR2021]  383572 non-null  object
 5   Topic          383572 non-null  object
dtypes: object(6)
memory usage: 17.6+ MB


In [8]:
#verificado baixa taxa de valores nulos, nesse caso, irei re importar os dados ja tratando os valores nulos

df = pd.read_excel('datasets/WDI World Bank.xlsx',
                  na_values='...')#substituindo os valores NaN por '...'

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 383577 entries, 0 to 383576
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   Country Name   383574 non-null  object
 1   Country Code   383572 non-null  object
 2   Series Name    383572 non-null  object
 3   Series Code    383572 non-null  object
 4   2021 [YR2021]  383572 non-null  object
 5   Topic          383572 non-null  object
dtypes: object(6)
memory usage: 17.6+ MB


In [9]:
#Simplificando e removedo espaços dos nomes das variaveis

df.rename(
    columns={
        'Country Name':'pais',
        'Country Code':'cod_pais',
        'Series Name':'serie',
        'Series Code':'cod_serie',
        '2021 [YR2021]':'ano_2021',
        'Topic':'topico'
    },
inplace = True
)

#Verificando se os nomes de fato foram trocados
df.columns

Index(['pais', 'cod_pais', 'serie', 'cod_serie', 'ano_2021', 'topico'], dtype='object')

In [10]:
serie_unique_topico = pd.Series(df['topico'].unique())
serie_unique_topico.dropna(inplace = True)
serie_unique_topico_health = serie_unique_topico[serie_unique_topico.str.startswith('Health')]
serie_unique_topico_health

5           Health: Reproductive health
6                  Health: Risk factors
7          Health: Population: Dynamics
22           Health: Disease prevention
38               Health: Health systems
42                    Health: Nutrition
68    Health: Universal Health Coverage
73                    Health: Mortality
79        Health: Population: Structure
82                               Health
dtype: object

In [11]:
#Utilizando minha seleção, realizarei um agrupamento com base em topico, onde topico se inicia em 'health'

pd.pivot_table(
    df[df['topico'].isin(serie_unique_topico_health)],
    index=['pais','cod_pais'],
    columns=['topico'],
    aggfunc='size'
)

topico                                Health  Health: Disease prevention  \
pais                        cod_pais                                       
Afghanistan                 AFG            2                          27   
Africa Eastern and Southern AFE            2                          27   
Africa Western and Central  AFW            2                          27   
Albania                     ALB            2                          27   
Algeria                     DZA            2                          27   
...                                      ...                         ...   
West Bank and Gaza          PSE            2                          27   
World                       WLD            2                          27   
Yemen, Rep.                 YEM            2                          27   
Zambia                      ZMB            2                          27   
Zimbabwe                    ZWE            2                          27   

topico                                Health: Health systems  \
pais                        cod_pais                           
Afghanistan                 AFG                           23   
Africa Eastern and Southern AFE                           23   
Africa Western and Central  AFW                           23   
Albania                     ALB                           23   
Algeria                     DZA                           23   
...                                                      ...   
West Bank and Gaza          PSE                           23   
World                       WLD                           23   
Yemen, Rep.                 YEM                           23   
Zambia                      ZMB                           23   
Zimbabwe                    ZWE                           23   

topico                                Health: Mortality  Health: Nutrition  \
pais                        cod_pais                                         
Afghanistan                 AFG                      39                 26   
Africa Eastern and Southern AFE                      39                 26   
Africa Western and Central  AFW                      39                 26   
Albania                     ALB                      39                 26   
Algeria                     DZA                      39                 26   
...                                                 ...                ...   
West Bank and Gaza          PSE                      39                 26   
World                       WLD                      39                 26   
Yemen, Rep.                 YEM                      39                 26   
Zambia                      ZMB                      39                 26   
Zimbabwe                    ZWE                      39                 26   

topico                                Health: Population: Dynamics  \
pais                        cod_pais                                 
Afghanistan                 AFG                                 13   
Africa Eastern and Southern AFE                                 13   
Africa Western and Central  AFW                                 13   
Albania                     ALB                                 13   
Algeria                     DZA                                 13   
...                                                            ...   
West Bank and Gaza          PSE                                 13   
World                       WLD                                 13   
Yemen, Rep.                 YEM                                 13   
Zambia                      ZMB                                 13   
Zimbabwe                    ZWE                                 13   

topico                                Health: Population: Structure  \
pais                        cod_pais                                  
Afghanistan                 AFG                                  58   
Africa Eastern and Southern AFE        

In [36]:
#Uma segunda maneira de obter um resultado semelhante 

df[df['topico'].isin(serie_unique_topico_health)].groupby(['pais','cod_pais','topico']).count()

serie  cod_serie  \
pais        cod_pais topico                                                
Afghanistan AFG      Health                                 2          2   
                     Health: Disease prevention            27         27   
                     Health: Health systems                23         23   
                     Health: Mortality                     39         39   
                     Health: Nutrition                     26         26   
...                                                       ...        ...   
Zimbabwe    ZWE      Health: Population: Dynamics          13         13   
                     Health: Population: Structure         58         58   
                     Health: Reproductive health           16         16   
                     Health: Risk factors                  32         32   
                     Health: Universal Health Coverage     13         13   

                                                        ano_2021  
pais        cod_pais topico                                       
Afghanistan AFG      Health                                    2  
                     Health: Disease prevention               27  
                     Health: Health systems                   23  
                     Health: Mortality                        39  
                     Health: Nutrition                        26  
...                                                          ...  
Zimbabwe    ZWE      Health: Population: Dynamics             13  
                     Health: Population: Structure            58  
                     Health: Reproductive health              16  
                     Health: Risk factors                     32  
                     Health: Universal Health Coverage        13  

[2660 rows x 3 columns]

In [34]:
#Para validar se os dados acimas estao corretos, para Afghanistan, temos 27 oberções onde o tóico era 'Health: Disease prevention'

#filtro tópico
health = df[df['topico'].isin(serie_unique_topico_health)]

#filtro pais
health = health[ health['pais'] == 'Afghanistan']

#Contagem
len(health[health['topico'] == 'Health: Disease prevention'])

#Verifiquei que o valor 27 é exatamente o valor que aparece para o tópico 'Health: Disease prevention' no dataframe anterior

27

In [13]:
#Verificando a quantidade de paises
len(df['pais'].unique())

269

269 paises parece um valor incorreto, visto que nao existe essa quantidade de paises no mundo